lib/gems/pending/util/mount/miq_s3_session.rb

```diff
@@ -0,0 +1,66 @@
+require 'util/mount/miq_generic_mount_session'
+require 'aws-sdk'
+
+class MiqS3Session < MiqGenericMountSession
+  def initialize(log_settings)
+    log_header = "MIQ(#{self.class.name}-initialize)"
+    logger.debug("#{log_header} initialize: log_settings are #{log_settings}")
+    super(log_settings)
+    logger.debug("#{log_header} initialize: @settings are #{@settings}")
+    raise "username is a required value!" if @settings[:username].nil?
+    raise "password is a required value!" if @settings[:password].nil?
+    raise "region is a required value!" if @settings[:region].nil?
+    @s3 = s3
+  end
+
+  def s3
+    @s3 ||= Aws::S3::Resource.new(:region => @settings[:region], :access_key_id => @settings[:username], :secret_access_key => @settings[:password])
+  end
+
+  def connect
+    @host       = URI(uri).host
+    @mount_path = URI(uri).path
+    super
+  end
+
+  def disconnect
+    return if @mnt_point.nil?
+    FileUtils.rm_rf(@mnt_point) if File.exist?(@mnt_point)
+  end
+
+  def uri_to_local_path(remote_file)
+    # Strip off the leading "s3:/" from the URI"
+    File.join(@mnt_point, URI(remote_file).host, URI(remote_file).path)
+  end
+
+  def uri_to_object_path(remote_file)
+    # Strip off the leading "s3://" and the bucket name from the URI"
+    # Also remove the leading delimiter.
+    URI(remote_file).path[1..-1]
+  end
+
+  def add(local_file, uri, object_file)
```

Comment:
I'm not sure we need an additional parameter here.

The original implementation of `#add` uses the filename in the destination uri (second parameter) as the remote file name. Is there any reason we couldn't do the same here?

ref: https://github.com/ManageIQ/manageiq-gems-pending/pull/355/files#diff-633c63845bfb2f447bd88a0ff73bacfdR256

Reply:
I'll take a look.

Reply:
The difference is that the S3 bucketname would be the first part of the URI, but I'm also using the rest of the URI as the "directory-like" key/prefix to the the filename.  And I'm doing the same thing with the filename that is done for Samba and NFS - using the file path that is created by the actual database dump.  So for instance, if we were using NFS, the file would get copied into a directory tree named "db_backup/region_0/__adhoc_dbbackup_2018_07_06_13_41_23_0400"  underneath the NFS mount point.

Reply:
I'm not sure I understand this. In the PR which uses this class it looks like `object_file` here just ends up being the result of `#uri_to_object_file` [here](https://github.com/ManageIQ/manageiq/pull/17689#discussion_r202808292). In my mind you could simplify this by removing this parameter and extracting all the information you need from the uri inside the processing done in this class.